---
title: "Code Library"
author: "Ruhika Chatterjee"
date: "2024-12-07"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Notes taken from Johns Hopkins University Coursera course series Data Science Specialization.

# R Data Types
- Basic object is vector of same class except list.  
- Atomic classes of objects: character, numeric (real), integer, complex, logical.  
- Attributes can include names, dimnames, dimensions, class, length.

### Atomic Data
``` {r atomicClass}

# numeric Vector
x <- 5 # numeric vector of 1 element

# integer vector
x <- 5L # integer vector of len 1

x <- Inf # special number infinity, +/-
x <- NaN # special number undefined, usually hijacks operations

# character vector
msg <- "hello" # char vector of len 1

# logical vector
tf <- TRUE # logical vector of value true
# TRUE = 1 = T, FALSE = 0 = F, num > 0 = TRUE

# complex vector
x <- 1+4i # vector of complex num of len 1

```

### complex Data Types
``` {r complexData}
# vector
vector("numeric", length = 10) # create vector of one type, args: class, length
c(1,2,3,4) # creates vector of common denominator class with given values
1:20 # vector sequence of 20 elements 1-20
pi:10 # will not exceed 10, start from pi, increment by 1
15:1 # increment -1
seq(1,20) # same as :
seq(0,10,by=0.5) # to change increment
seq(5,10,length=30) # to not set increment but number of numbers
seq_along(x) # vector of same length 1:length(x)
rep(10, times = 4) # repeats 10 4 times in vector
rep(c(0, 1, 2), times = 10) # repeats sequence of vector 10 times. Arg each can be used to repeat first, then second... element in vector.

# lists
# vector capable of carrying different classes
x <- list(1, "a", TRUE, 1+4i) # vector of vectors

# Matrix
# vector of single class with rectangular dimensions (attribute of integer vector len 2)
x <- matrix(nrow=2,ncol=3) # empty matrix of given dimensions
x <- matrix(1:8, nrow = 4, ncol = 2) # creates matrix of given dimensions with values assigned, created by column
y <- matrix(rep(10,4),2,2) # creates matrix of 4 10s

x <- 1:10
dim(x) <- c(2,5) # creates matrix out of vector with dimension 2 rows x 5 columns

cbind(1:3,10:12) # creates matrix out of values in vector args, adding by column (1st arg = 1st col)
rbind(1:3,10:12) # same but using rows

# factors
# self-describing type of vector representing categorical data, ordered or unordered (labels)
x <- factor(c("male","female","female","female","male")) # character vector with specific linear modeling capabilities, levels also in output
f <- gl(3,10) # factor 3 levels, 10 times each
table(x) # prints counts of each factor

```

### Data Frames
``` {r dataframe}
# stores tabular/rectangular data, stored as lists of same length where each element is a column, length of element is number of rows. Different classes possible
x <- data.frame(foo=1:4, bar=c(T,T,F,F)) # creates data frame 2 columns foo and bar, 4 rows unnamed. Can also combine with matrices
x <- read.table(file = "hw1_data.csv", header = TRUE, sep = ",") # read in data from file
x <- read.csv("hw1_data.csv") # same

```

### Date and Time Data Types
``` {r datetime}
# useful for time-series data (temporal changes) or other temporal info
# lubridate package by Hadley Wickham

# Dates and Times
birthday <- as.Date("1970-01-01") # dates are date class defined by converting character string, year-month-day, stored as number of days since 1970-01-01 (previous is negative)
today <- Sys.Date()

currentTime <- Sys.time()# time by POSIXct(large integer vector, useful in dataframe) or POSIXlt(list, other info stored like day) class, stored internally as number of seconds since 1970-01-01 or list, respectively
timedefined <- as.POSIXct("2012-10-25 06:00:00") # convert char vector, can define timezone
cTConvert <- as.POSIXlt(currentTime) # reclass, works other way
cTConvert$min # to subset list

datestring <- c("January 10, 2012 10:40", "December 9, 2011 9:10")
x <- strptime(datestring, "%B %d, %Y %H:%M") # Convert character vector to POSIXlt by defining format (format string)
x

weekdays(birthday) # return day of week, date or time classes
months(birthday) # return month on date or time
quarters(birthday) # return quarter of date or time

# Operations
# CANNOT MIX CLASSES - convert
# add and subtract dates, compare dates
currentTime - timedefined # time difference, track of discrepancies (i.e. daylightsavings, timezones, leap time, etc). Can do + and - or <, >, <=, >=, ==, !=.
difftime(currentTime, timedefined, units = "days") # to specify unit

```


# Basic R Functions

### Functions and Operations
``` {r basic}
# Input and Evaluation
x <- 1 # assignment operator, evaluates and returns
print(x) # print value as vector
x # auto-prints
# in console, press Tab for auto-completion
LETTERS # predefined character vector of capital letters
# <<- operator can be used to assign a value to an object in an environment that is different from the current environment

# Mathematical and Statistical Functions
5 + 7 # basic arithmetic operations all work +, -, *, /, ^, %% (modulus). NA affects operation.
sqrt(4) # square root
abs(-1:2) # absolute value
mean(c(3,4,5,6,7)) # return mean of numeric vector
sd(c(3,4,5,6,7)) # returns standard deviation of numeric vector
cor(c(3,4,5,6,7), c(61,47,18,18,5)) # correlation of x and y vectors make sure to set arg use for NAs
range(c(3,4,5,6,7)) # returns min and max as numeric vector of 2
quantile(c(3,4,5,6,7), probs = 0.25) # returns 25th percentile
-c(0.5,0.8,10) # distributes the negative to all elements of vector

# vectorized operations
x <- 1:4; y <- 6:9 # different length vectors
x + y # add the elements of the vectors, all operators work
x > 2 # returns logical vector, >= or == or any of the logical expressions work

# Matrix Operations
x <- matrix(1:4,2,2); y <- matrix(rep(10,4),2,2)
x * y # element wise multiplication, for all operators
x %*% y # matrix multiplication

x <- matrix(rnorm(200), 20, 10)
rowSums(x) # vector of sum of rows
rowMeans(x) # vector of mean of rows
colSums(x) # vector of sum of cols
colMeans(x) # vector of mean of cols
x <- matrix(rnorm(100), 10, 10)
solve(x) # returns inverse of matrix if invertible

# Logical operators
5 >= 2 # returns logical. <, >, <=, >=, ==, !=. NA in expression returns NA. Can also use to compare logical vectors themselves.
TRUE | FALSE # OR A|B union, AND A&B intersection, NOT !A negation. & operates across vector, && evaluates first element of vector. Order of operations: AND before OR
isTRUE(6 > 4) # also evaluates logical expression
xor(5 == 6, !FALSE) # only returns TRUE if one is TRUE, one is FALSE
which(c(1,2,3,4,5,6) < 2) # returns indices of logical vector where element is TRUE
any(c(1,2,3,4,5,6) < 2) # returns TRUE if any of the logical index values are TRUE
all(c(1,2,3,4,5,6) < 2) # returns TRUE only if all the elements of vector are TRUE

# Character functions
paste(c("My","name","is"),collapse = " ") # join elements into one element, can join multiple vectors with multiple arguments, adds element-by-element with coersion and recycling
c (c("My","name","is"), "Bob") # add to the vector

# Factors functions
x <- factor(c("male","female","female","female","male")) # can include levels argument to set order (baseline is first) otherwise alphabetical
x # prints values in vector and levels
table(x) # prints labels and counts present
unclass(x) # strips class to integer with levels of labels

# Display Data Functions
print(data.frame(foo = 1:20, rar = 301:320)) # print whole object
head(data.frame(foo = 1:20, rar = 301:320)) # prints preview of first 6 lines
tail(data.frame(foo = 1:20, rar = 301:320)) # prints preview of last 6 lines
table(c(1,1,1,2,2,2,2,2,2,2,2,2,3,3,3,3,4,4,5)) # returns table of counts
summary(c(3,4,5,6,7)) # result summaries of the results of various model fitting functions based on class
unique(c(3,4,5,6,7,3,3,5,7,2,8,3,5,6)) # returns only unique elements, duplicates removed

# str function - compactly display internal structure of R object (esp large lists). Diagnostic, alternative to summary()
str(unclass(as.POSIXlt(Sys.time()))) # prints list clearly
str(lm) # list of function arguments
str(rnorm(100,2,4)) # type of vector, length, first 5 elements
str(gl(40,10)) # for factors

# Missing Values
# represented as NA (missing, with specified class) or NaN (missing or undefined)
# NaN is NA but NA not always NaN
is.na(c (1,2,NA,5,6,NA, NA,3, NaN)) # output logical vector of length of input
is.nan(c (1,2,NaN,5,6,NA, NaN,3)) # output logical vector of length of input

```


### Attributes of objects
``` {r attributes}
x <- c(0.5,105,10,0.1,2)
class(x) # determine class of object
attributes(x) # function to return or modify attributes of object
identical(x,x) # returns logical for if two objects are identical
length(x) # to specifically get the length of vector
dim(x) # to get dimensions of matrix, data frame (row, column)
object.size(x) # return memory occupied in bytes
as.numeric(0:6) # explicit coercion, works on all atomic classes, if not possible converts to NA and warning

# data frames
row.names(x) # get and set row names (attributes). Can also use rownames(x)
colnames(x) # get and set row names
nrow(x) # number of rows
ncol(x) # number of columns
data.matrix(x) # converts data frame to matrix, coercion
dim(x) # (row, column) dimensions of data frame

# names attribute
x <- 1:3
names(x) # is null
names(x) <- c("foo","bar","norf") #now not numbered vector but named, print x and names(x) with names
vect <- c(foo = 11, bar = 2, norf = NA) # adds elements with names to vector directly
# also for lists, names vectors not items
m <- matrix(1:4,nrow = 2, ncol = 2)
dimnames(m) <- list(c("a","b"),c("c","d")) # each dimension has a name for matrices, rows names then columns

```

### Indexing, Subsetting, and Dealing with NAs
``` {r indexing}
# Subsetting R Objects
x <- c("a","b","c","c","d","a")
x[1] # more than one element extracted, returns same class as the original, numeric/logical index
x[1:4] # sequence of num index
x[x>"a"] # logical indexing, returns vector where logical is true
u <- x > "a" # create logical vector
x[u] # same as x[x>"a"]
x[!is.na(x) & x > 0] # returns only positive, non NA values
x[c(-2, -10)] # returns vector with 2nd and 10th elements removed
x <- data.frame(foo = 1:6, bar = c("g","h","i","j","k","l"))
x[[which(x$bar == "h"), "foo"]] # get or set foo in the same row as bar of "h"

x <- list(foo = 1:4, bar = 0.6, baz = "hello")
x[1] # list containing first element
x[[1]] # extract from list/data frame, single element, class can change. Ex, numerical vector returned
x$bar # like [[]] but by name. Ex, return num vector 0.6. Equivalent to x[["bar"]]. Expression x["bar"] returns list of bar. Can use partial matching (x$f = x$foo = x[["f", exact = FALSE]]).
x[c(1,3)] # multiple object extraction from list, returns list
name = "foo"
x[[name]] # must be used if using computed index
x[1][3] # return element in element in object
x[[c(1,3)]]

# Subsetting Matrix
x <- matrix(1:6, 2, 3)
x [1,2] # returns vector len 1, different that x[2,1]. Get matrix using arg drop = FALSE.
x[1,] # get num vector of first row, can also get col x[,2]. drop = FALSE also works

# Removing NA values
x <- c(1,2,NA,4,NA,5)
bad <- is.na(x) # logical vector indicating presence of NA
x[!bad] # removes NA values

x[!is.na(x)] # simplified returns vector removing NA values

x <- c(1,2,NA,4,NA,5) # for two vectors
y <- c("a","b",NA,"d",NA,"f")
good <- complete.cases(x,y) # logical vectors where there is no NA in either list
x[good]
y[good]

# Sum of NA values
my_na <- is.na(x)
sum(my_na)

x <- read.csv("hw1_data.csv") # for data frames
goodVals <- complete.cases(x) # complete rows in the data frame
x[goodVals,]

```

### Data Tables (not Frames)
- Package, faster and more memory efficient  
- Inherets from data.frame (all functions), written in C, faster at sub-setting, grouping, and updating  
- <http://stackoverflow.com/questions/13618488/what-you-can-do-with-data-frame-that-you-cant-in-data-table>  
- <https://github.com/Rdatatable/data.table>  

``` {r dataTable}
library(data.table)
DF = data.frame(x=rnorm(9),y=rep(c("a","b","c"), each=3),z=rnorm(9))
head(DF,3)

DT = data.table(x=rnorm(9),y=rep(c("a","b","c"), each=3),z=rnorm(9))
head(DT,3)
tables() # get all data tables in memory

# subsetting
DT[2,] # subset rows
DT[DT$y=="a",] # subset where y is "a"
DT[c(2,3)] # subset rows 2 & 3, one variable is assigned to rows
# subset cols, DT[,c(2.3)] does not work bc uses expressions
DT[,list(mean(x),sum(z))] # pass list of functions applied by names of columns
DT[,table(y)] # get table of y values
DT[, w := z^2] # adds columns quickly
DT2 <- DT # does not make a copy in memory, change one changes all, pointing to same memory. Use copy function
DT[,m:= {tmp <- (x+z); log2(tmp+5)}] # multiple step function, returns last statement in evaluation
DT[,a:=x>0] # expression exaluates boolean for new variable
DT[,b:= mean(x+w),by=a] # grouping by boolean a into factors to evaluate expression
# special variable .N integer len 1 num times group appears
set.seed(123)
DT <- data.table(x=sample(letters[1:3], 1E5, TRUE))
DT[, .N, by=x] # count number of times grouped by x variable

# data.table contains keys
DT <- data.table(x=rep(c("a","b","c"),each=100), y=rnorm(300))
setkey(DT, x)
DT["a"] # subset based on key x, faster
DT1 <- data.table(x=c("a","a","b","dt1"), y=1:4)
DT2 <- data.table(x=c("a","b","dt2"), z=5:7)
setkey(DT1,x); setkey(DT2,x)
merge(DT1,DT2) # uses keys to merge

# fast reading in data.table
big_df <- data.frame(x=rnorm(1E6),y=rnorm(1E6))
file <- tempfile()
write.table(big_df, file=file, row.names=FALSE, col.names=TRUE, sep="\t", quote=FALSE)
system.time(fread(file)) # basically read.table for csv
system.time(read.table(file,header=TRUE,sep="\t"))

```


# Random Numbers
``` {r randomnumbers}
# Random number generation
# Probability distribution functions have 4 functions associated: d- density, r- random number generation, p- cumulative distribution, q- quantile function. Normal, gamma, Poisson, binomial, exponential, chi-squared
set.seed(1) # set sequence of random number generation. set.seed(1); rnorm(5) always results in the same sequence (reproducible).
y <- rnorm(1000) # generate vector of 1000 numbers that are standard normal distribution. Agrs: n, mean = 0, sd = 1.
y <- dnorm(c(0.25,0.5,0.75)) # evaluate Normal probability density, (given mean,sd) at point or vector points. Args: x, mean = 0, sd = 1, log = FALSE.
y <- pnorm(0.5) # evaluate cumulative distribution function for normal distribution. Args: q, mean=0, sd=1, lower.tail=TRUE, log.p=FALSE.
y <- qnorm(0.5) # evaluates quantiles for normal distribution. Args: p, mean=0, sd=1, lower.tail=TRUE, log.p=FALSE.
y <- sample(1:6,3) # random selection of 3 elements from array
ints <- sample(10) # random sample all integers from 1 to 10 without replacement. Permutation
nums <- sample(1:10, replace = TRUE) # with replacement
let <- sample(LETTERS) # sample all letters without replacement
flips <- sample(c(0,1), 100, replace = TRUE, prob = c(0.3,0.7)) # unfair coin
coin <- rbinom(1,1,0.5) # simulating coin flip
unfairflip <- rbinom(1, size = 100, prob = 0.7) # sum of flips above
flips2 <- rbinom(100,1,0.7) # flips above
y <- rpois(10, 1) # generate random poisson variates with given rate. Args: n (count), rate (mean)
pois_mat <- replicate(100, rpois(5, 10))

# Simulate Linear Model Ex
# y = B(o) + B(1) * x + e
# e ~ N(0,2^2) assume x ~ N(0,1^2), B(0) = 0.5, B(1) = 2.
set.seed(20)
x <- rnorm(100)
e <- rnorm(100,0,2)
y <- 0.5 + 2 * x + e
# can combine different distributions
# Poisson: Y ~ Poisson(mu)
# log(mu) = B(0) + B(1)x
# B(0) = 0.5 and B(1) = 0.3
set.seed(1)
x <- rnorm(100)
log.mu <- 0.5 + 0.3 * x
y <- rpois(100, exp(log.mu))

```


# Control Functions and Loop Functions

### Control Functions
``` {r controlfunctions}
# control execution of program

x = 2
# if,else loops
y <- if(x > 3){ # testing condition
  10
} else if(x > 0 & x <= 3) { # can not have or multiple
  5
} else{ # can not have, at end
  0
}

if(x-5 == 0){
  y <- 0
} else{
  y <- 2
}

# for loops
for(i in 1:10) {# execute loop fixed number of times. Args iterator variable and vector(inc seq) or list
  print(i)
}

x <- c("a","b","c","d")
for(i in 1:4){
  print(x[i])
}
for(i in seq_along(x)){
  print(x[i])
}
for(letter in x){
  print(letter)
}
for(i in 1:4) print(x[i])

x <- matrix(1:6,2,3)
for(i in seq_len(nrow(x))) { # nested, don't use more than 2-3 for readability
  for(j in seq_len(ncol(x))) {
    print(x[i,j])
  }
}

# while loops
count <- 0
while(count < 10){ # loop while condition is true
  print(count)
  count <- count + 1
} # be wary of infinite loops!! when condition cannot be true

z <- 5
while(z >= 3 & z <= 10){
  print(z)
  coin <- rbinom(1,1,0.5)
  
  if (coin == 1) z <- z+1
  else z <- z-1
}

# Repeat loop
x0 <- 0.01; tol <- 1e-3
repeat { # infinite loop
  x1 <- rnorm(1)
  if(abs(x1 - x0) < tol) {
    break # break execution of any loop
  }
  else x0 <- x1
}

# control a loop
for(i in 1:100) {
  if(i <= 20) next # skip next iteration of loop
  else {
    if (i > 50) break # exit for loop
  }
}

# return to exit a function, will end control structure inside function

```

### Loop Functions
``` {r loopfunctions}
# Loop functions - useful for looping in the command line
# Hadley Wickham's Journal of Statistical Software paper titled 'The Split-Apply-Combine Strategy for Data Analysis'.

# lapply - loop over a list and evaluate on each element. args: X (list or coercion), FUN (function or function name), ... (other args). returns list with same names if defined. Passes element to function first argument
x <- list(a = 1:5, b = rnorm(10))
lapply(x, mean) # returns list of 2 numerics
x <- 1:4
lapply(x, runif, min = 0, max = 10) # passes subsequent args to function
x <- list(a = matrix(1:4, 2, 2), b = matrix(1:6, 3, 2))
lapply(x, function(elt) elt[,1]) # define an anonymous function inside lapply

# sapply - same as lapply but simplify, i.e. will make list of 1 element vectors a vector, multiple elements same length returns a matrix, else a list
x <- list(a = 1:5, b = rnorm(10))
lapply(x, mean) # now returns vector length 2
# mean only operates on signle element numeric/logical, so need to use loop

# vapply - pre-specify type of return value, safer and faster. Args: X, FUN, FUN.VALUE (generalized vector as template for return), ..., USE.NAMES = TRUE)
vapply(x, mean, numeric(1)) # same as sapply(x, mean)

# apply - apply function over margins of array (good for summary of matrices or higher level array). Not better than for loop except less code. Args: X as above, MARGIN (integer vector showing margins to retain), FUN as above, ... as above
x <- matrix(rnorm(200), 20, 10)
apply(x, 2, mean) # mean of each column by collapsing 1st dimension, returns num vector length of ncol. MARGIN = 1 gives rows collapse cols.
rowSums(x) # equivalent to apply(x, 1, sum)
rowMeans(x) # equivalent to apply(x, 1, mean)
colSums(x) # apply(x, 2, sum)
colMeans(x) #apply(x, 2, mean)
apply(x, 1, quantile, probs = c(0.25, 0.75)) # runs quantile with 2 agrs for every element in list, returns matrix of 2 rows, nrows rows.
a <- array(rnorm(2 * 2 * 10), c(2, 2, 10)) # array in 3D
apply(a, c(1,2), mean) # collapses only 3rd dimension, returns 2x2 matrix. Equivalent rowMeans(a, dims = 2).

# tapply - apply function over subset of a vector. args: X is vector, INDEX is factor/list factors vector same length as X, FUN, ..., simplify = TRUE
x <- c(rnorm(10), runif(10), rnorm(10,1))
f <- gl(3,10) # factor 3 levels, 10 times each
tapply(x,f,mean)

# mapply - multivariate version of lapply. args: FUN as above, ... (arguments to apply over), MoreArgs (other arguments for FUN), SIMPLIFY (logical for if to simplify). Apply over multiple lists. Like nested loop
list(rep(1,4), rep(2,3), rep(3,2), rep(4,1))
mapply(rep, 1:4, 4:1) # equivalent

noise <- function(n,mean,sd){rnorm(n,mean,sd)}
noise(1:5,1:5,2) # gives vector of 5, same as single num args
mapply(noise,1:5,1:5,2) # applies function for each pair, list of 5 of length i

# split - in conjunction with lapply to split objects into subpieces. Args: x (any object), f (factor), drop (logical if empty factors to be dropped) = FALSE, ...
x <- c(rnorm(10), runif(10), rnorm(10,1))
f <- gl(3,10) # factor 3 levels, 10 times each
split(x,f) # tapply without function, sorts into list based on levels, can then use lapply or sapply.
lapply(split(x,f), mean) # in this case can use tapply
# can do data frames
data <- read.csv("hw1_data.csv")
s <- split(data, data$Month)
sapply(s, function(x) colMeans(x[,c("Ozone","Solar.R","Wind")], na.rm = TRUE)) # data$Month coerced into factor
# Multi-level split
x <- rnorm(10)
f1 <- gl(2,5); f2 <- gl(5,2) # ex. race and gender 2 factors
interaction(f1,f1) # combine each pair, 10 factors
split(x, list(f1,f2)) # interaction called, list returned for combination sort, drop = TRUE to remove unused

```

# Defining Functions
``` {r defFunctions}
# stored in txt or R script, functions are R objects. Can pass functions as arguments for other functions, functions can be nested.

myfunction <- function(){ #create a function
  x <- rnorm(100)
  mean(x)
}
myfunction() #call created function
myfunction # prints source code for function
args(myfunction) # returns arguments for passed function

myaddedfunction <- function(x,y){ #create a function with formal arguments x and y
  x + y + rnorm(100) # implicit return last expression
}
myaddedfunction(5,3)
myaddedfunction(4:10,2)

# function with default argument if left unspecified, for common cases
above <- function(x, n = 10){
  use <- x > n
  x[use]
}
above(1:20) # n is default set to 10
above(1:20, 12) # n set at 12

columnmean <- function(y, removeNA = TRUE) {
  nc <- ncol(y)
  means <- numeric(nc)
  for(i in 1:nc) means[i] <- mean(y[,i], na.rm = removeNA)
  invisible(means) # auto-return blocks auto-print
}

# Lazy Evaluation: R evaluated statements and arguments as they come
f <- function (a,b,c){
  print(a)
  #print(b) # error
}
f(3) # prints a, error for b, no rxn to not having c

# ways to call functions
# positional matching and naming can be mixed. Partial matching also allowed, if not found uses positional match.
# named helps for long arg list where most defaults are maintained or if order is hard to remember.
mydata <- rnorm(100)
sd(mydata) # default to first argument
sd(x = mydata)
sd(x = mydata, na.rm = FALSE)
sd(na.rm = FALSE, x = mydata)
sd(na.rm = FALSE, mydata) # remove argument from list, default works on first unspecified arg

# Variable Arguments
# to extend another function without copying arg list of OG function
simon_says <- function(...){
  paste("Simon says:", ...)
}
# or for generic functions passed to methods
# unpacking an ellipses
mad_libs <- function(...){
  args <- list(...)
  place <- args$place
  adjective <- args$adjective
  noun <- args$noun
  paste("News from", place, "today where", adjective, "students took to the streets in protest of the new", noun, "being installed on campus.")
}
# or when number of args unknown in advance (if at beginning, no positional or partial matching)
args(paste) # operates on unknown sets of character vectors

# function as an argument
some_function <- function(func){
  func(2, 4) # returns result of function with 2,4 arguments
}
some_function(mean) # returns mean of 2,4

# Anonymous function (chaos)
evaluate  <- function(func, dat){
  func(dat)
}
evaluate(function(x){x+1}, 6) # creates a function when calling evaluate to add 1

# create a binary operation
"%mult_add_one%" <- function(left, right){
  left * right + 1
}
4 %mult_add_one% 5

```

### Lexical Scoping
``` {r lexicalscoping}
make.power <- function(n) {
  pow <- function(x) {
    x^n
  }
  pow
}

cube <- make.power(3)
square <- make.power(2)
cube(3)
square(3)

# Scoping - environments
search()# provides list of environments
ls(environment(cube)) # object names in function environment, same for square
get("n",environment(cube)) # values in function environment, changes for square

```


# R Packages
- Repositories: CRAN, BioConductor (bioinformatics), GitHub  
- Search: https://www.rdocumentation.org/  

- Base packages: utils, stats, datasets, graphics, grDevices, grid, methods, tools, parallel, compiler, splines, tcltk, stats4.  
- Recommended packages: boot, class, cluster, codetools, foreign, KernSmooth, lattice, mgcv, nime, rpart, survival, MASS, spatial, nnet, Matrix.  
``` {r packages}
# Install from CRAN:
#   install.packages("ggplot2", repos = "http://cran.us.r-project.org") #install
#   install.packages(c("labeling","tibble"), repos = "http://cran.us.r-project.org") #multiple

# Install from Bioconductor
#   install.packages("BiocManager", repos = "https://bioconductor.org/biocLite.R")
#   BiocManager::install(c("GenomicFeatures", "AnnotationDbi")) #install package

# Install from GitHub (need package, author name)
#   install.packages("devtools", repos = "http://cran.us.r-project.org") #only once
#   library(devtools)
#   install_github("author/package") #installs package

# library(ggplot2)# Load package, careful of dependencies
# installed.packages() #check installed packages
# library() #alternate
# old.packages(repos = "http://cran.us.r-project.org") #check packages to update
# update.packages(repos = "http://cran.us.r-project.org") #update all packages
# install.packages("ggplot2") #to update single package
# detach("package:ggplot2", unload=TRUE) #unload function
# remove.packages("ggtree") #remove package
# help(package = "ggplot2") #package info
# browseVignettes("ggplot2") #extended help files

```


# Getting the Data

### Defining Workspace or directory
``` {r dirws}
x <- getwd() # find working directory
x
dir.create("testdir") # create a directory if doesn't exist, args: dir name, for nested recursive = true
setwd("testdir") # set working dir
file.create("mytest.R") # create file in wd
file.exists("mytest.R") # check if file or directory exists in wd
file.info("mytest.R") # file metadata, use $ operator to grab specific items
file.rename("mytest.R","mytest2.R") # rename
file.copy("mytest2.R","mytest3.R") # copy file
file.remove("mytest2.R") # remove file
file.path("mytest3.R") # relative path
setwd(x) # could use relative setwd("../") to move up one, setwd("./data"), or absolute path directly using setwd("C:\\Users\\Owner\\OneDrive\\Documents")

dir() # output files in directory. Also list.files()
files_full <- list.files("specdata", full.names=TRUE) # pull all file names from a directory
ls() # prints the objects in work space
rm(list=ls()) # clear workspace
rm(list=setdiff(ls(), "x")) # clear workspace except x
version #R info version
sessionInfo() #R info version, packages
source("coded.R") # load code into console
args(ls()) # get arguments for a function
help(ls) # access documentation on ls() function
?ls # same. for operator use ?`:`

# Interface to outside world
file(description = "hw1_data.csv") # open connection to standard, uncompressed file. Helps for partial file reading. description = file name, open = "r"(read only), "w" (writing and initializing new file), "a" (appending), "rb"/"wb"/"ab" in binary (Windows)
# gzfile() # connection to file w compression gzip
# bzfile() # connection to file w compression bzip2
jh <- url("http://www.jhsph.edu", "r") # connection to webpage
close(jh) # to end connection

```

### Reading and Writing Data
``` {r readingandwriting}
# Download data from website
fileUrl <- "https://stg-arcgisazurecdataprod1.az.arcgis.com/exportfiles-1506-24014/Part1_Crime_Beta_5960161298247612570.csv?sv=2018-03-28&sr=b&sig=8KeSXLduFH5b%2Bs4r30dNZ%2FIxL4qPiHHfi4XjwmFZX0I%3D&se=2025-01-06T02%3A30%3A50Z&sp=r"
if(!file.exists("./testdir/crimedata.csv")) download.file(fileUrl, destfile = "./testdir/crimedata.csv", method = "curl") # improve reproducibility. Args: url, destfile, method. curl method is for https (specifically MAC).
dateDownloaded <- date() # Keep track of date downloaded for file

# Read table or csv data into R
x <- read.table("hw1_data.csv", header = TRUE, sep = ",") # reading tabular data from text files, return data frame, reads into RAM (best for smaller data). Args: file (name, connection), header (logical if header line present), sep (string indicating column separator), colClasses (character vector of class of each column), nrows (number of rows), comment.char (character string indicating comment character), skip (number of lines to skip from beginning), stringsAsFactors (logical if character variables coded as factors, default true), na.strings (set char represent missing value). Efficient. Default separator " " and header = FALSE. Check help page to optimize for large datasets, set comment.char = "" to optimize if no comments present, use colClasses, nrow (memory).
head(x)
x <- read.csv("hw1_data.csv") # Same but default separator is ", "  and header = TRUE
# write.table(x)

# help read.table with colClasses with smaller sample
initial <- read.table("hw1_data.csv", header = TRUE, sep = ",", nrows = 100)
classes <- sapply(initial, class)
tabAll <- read.table("hw1_data.csv", header = TRUE, sep = ",", colClasses = classes)

# Read Excel file into R
x <- "https://stg-arcgisazurecdataprod1.az.arcgis.com/exportfiles-1506-24014/Part1_Crime_Beta_3834807016258906358.xlsx?sv=2018-03-28&sr=b&sig=47D%2FVGpnXbKodvr50TpJ40oOrvovymZfj2zHvDNVjYg%3D&se=2025-01-06T02%3A36%3A24Z&sp=r"
if(!file.exists("./testdir/crimexl.xlsx")) {
  download.file(x, destfile = "./testdir/crimexl.xlsx", method = "curl") # improve
  dateXl <- date()
}
library(readxl) # xlsx, XLConnect other options
cameraData <- read_excel("./testdir/crimexl.xlsx", sheet = 1, col_types = "text", n_max = 100)
head(cameraData)
# write also works

lines <- readLines("coded.R") # reading lines of text file, return character vector
writeLines("coded.R")

# editable textual format retains metadata, helpful for version control, corruption fixable, memory cost
dget("coded.R") # reading R objects deparsed into text files
dput("coded.R") # takes R object, create R code to reconstruct object saving attributes, names
source("coded.R") # reading in R code files
#dump() # multiple R objects

# load() # read in saved workspace read binary objects into R
# save()
# unserialize() # read single R objects in binary form
# serialize()

```

### Just XML stuff
- XML (extensible markup language)  
- <http://en.wikipedia.org/wiki/XML>  
- store structured data in internet applications. From web scraping, API  
- Components: markup (labels for structure) and content (text)  
- Tags (labels): start <section> end </section> empty <line-break />  
- Elements: example of tag <Greeting> Hello, world </Greeting>  
- Attributes: component of label <img src="jeff.jpg" alt="instructor"/>  

- XPath language: <http://www.stat.berkeley.edu/~statcur/Workshop2/Presentations/XML.pdf>  
- /node (top level node), //node (node any level), node[@attr-name] node w attr name, node[@attr-name='bob']
- XML Short Intro: <https://www.omegahat.net/RSXML/shortIntro.pdf>  
- XML Long Intro: <https://www.omegahat.net/RSXML/Tour.pdf>

``` {r xml}
# Reading file
library(XML)
library(httr)
fileUrl <- "https://www.w3schools.com/xml/simple.xml"
doc <- xmlTreeParse(rawToChar(GET(fileUrl)$content),useInternalNodes = TRUE) # read file into R to parse
rootNode <- xmlRoot(doc) # wrapper for structured element

# Access Data
xmlName(rootNode) # name of xml
names(rootNode) # names of root node
rootNode[[1]] # first food element, get first subelement of element using [[1]][[1]]
xmlSApply(rootNode,xmlValue) # give xml value of each root node
xpathSApply(rootNode,"//name",xmlValue) # get node xmlValue for each names
xpathSApply(rootNode,"//price",xmlValue)

# Pokemon
library(rvest)
fileUrlBase <- "https://pokemondb.net/pokedex/stats/gen"
generations <- 1:8
# Scrape Data
pokemonList <- lapply(generations,function(x){
  html <- read_html(paste0(fileUrlBase,x)) # create URL
  tableNode <- html_node(html,xpath="//*[(@id = 'pokedex')]") # read HTML, extract node using xpath selector
  data <- html_table(html,header=TRUE)[[1]] # convert to data frame
  colnames(data)[1] <- "ID"
  data$Generation <- x # add generation ID
  data # return data frame to list
})
thePokemon <- do.call(rbind,pokemonList) # one data frame
# Cleaning
types <- strsplit(gsub('[a-z]\\K(?=[A-Z])', ' ',thePokemon$Type, perl=T)," ") # split Type column
thePokemon$Type1 <- unlist(lapply(types,function(x) x[1])) # Type1 column
thePokemon$Type2 <- unlist(lapply(types,function(x) ifelse(is.na(x[2])," ",x[2]))) # Type2 column
pokemonNames <- strsplit(gsub('[a-z]\\K(?=[A-Z])', '--', thePokemon$Name, perl=T),"--") # split Name
thePokemon$Name <- unlist(lapply(pokemonNames,function(x) x[1]))
thePokemon$Form <- unlist(lapply(pokemonNames,function(x) ifelse(is.na(x[2])," ",x[2])))
thePokemon <- thePokemon[,c(1,2,14,12,13,4,5,6,7,8,9,10,11)] # reorder and remove unneccesary
thePokemon[29,"Name"] <- "Nidoran" # cleaning gender
thePokemon[29,"Form"] <- "Female"
thePokemon[32,"Name"] <- "Nidoran"
thePokemon[32,"Form"] <- "Male"

```

### Just JSON Stuff
- Javascript Object Notation- light weight data storage, common format from APIs, similar to XML but different syntax/format.  
- Data stored as: number(double), strings(double quoted), boolean, array(ordered, comma sep, [] enclosed), objects(unordered, comma sep, {} enclosed, key:value pairs).  
- <http://en.wikipedia.org/wiki/JSON>  
- <http://www.json.org/>  
- jsonlite vignette  
- <http://www.r-bloggers.com/new-package-jsonlite-a-smarter-json-encoderdecoder/>

``` {r json}
library(jsonlite)
jsonData <- fromJSON("https://api.github.com/users/jtleek/repos") # can pass any format
names(jsonData) # names of data frame
names(jsonData$owner) # names of column
jsonData$owner$login # who logged in

# write data frame to JSON
myjson <- toJSON(iris, pretty = TRUE)
cat(myjson)

# upload again
iris2 <- fromJSON(myjson)
head(iris2)

```

### mySQL Stuff
- 

``` {r sql}

```


# Cleaning Data
- End of process generate: raw data, tidy data set, code book (metadata) describing each variable and its values in the tidy data set, explicit and exact recipe used to convert raw data to tidy data set and code book.  
- Raw Data: original source of data i.e. no processing, editing, summarizing. Must process for data analysis (merging, sub-setting, transforming, etc.), be mindful of processing standards. Colloquially may be later step i.e. genome seq but must use rawest.  
- Processed Data: ready for analysis, steps to reach stage must be recorded. Must: one variable per column, each observation in a separate row, different tables for different types of variables, if multiple tables allow for linking. Useful: top row of variable names which are human readable, save one file per table.   
- Code book: info about variables not contained in data incl. units called *Code book*, info about summary choices, info about experimental study design called *Study design*. Often word/text file.  
- Instruction list: in a computer script where input is raw data and output is tidy data with no parameters to the script. If not possible, provide instructions in steps (incl. parameters, software versions, how to use software).  

# Graphics
<http://www.ling.upenn.edu/~joseff/rstudy/week4.html>
``` {r basegraphs}
# Start by getting sense of the data: dim(), names(), head(), tail() and summary().

# Scatterplot
data(cars)
plot(cars) # generates scatterplot with two columns against each other
plot(x = cars$speed, y = cars$dist) # explicit, can also pass plot(dist ~ speed, cars). Args: x, y = NULL, type = "p",  xlim = NULL, ylim = NULL, log = "", main = NULL, sub = NULL, xlab = NULL, ylab = NULL, ann = par("ann"), axes = TRUE, frame.plot = axes, panel.first = NULL, panel.last = NULL, asp = NA, xgap.axis = NA, ygap.axis = NA, ... (more at ?par)

# Boxplot
data(mtcars)
boxplot(mpg ~ cyl, data = mtcars) # generate relationship between mpg(x) and cyl(y) from mtcars

# Histogram
hist(mtcars$mpg) # generate a histogram of vector

```

# R Profiler and Optimization
- Systematic way to examine time spent in various part of the program. Useful to optimize the code.  
- DON'T PREMATURELY OPTIMIZE  
- Measure, not guess, data on what needs to be optimized.  
- User time: computer experienced, may be greater if multiple cores/processors (accessible in multi-threaded BLAS libraries). Elapsed time: wall-clock time, may be greater if other computing tasks.  
``` {r profiler}
system.time(read.csv("hw1_data.csv")) # returns seconds to execute, if error then seconds to error. Wrap longer expression in arguement in curly braces. Assumption htat you know where the problem is.

data(mtcars)
Rprof() # track function call stack at intervals (def = 0.02 sec), time spent in functions.

fit <- lm(mtcars$mpg ~ mtcars$cyl)

Rprof(NULL)

summaryRprof() # makes Rprof readable, tabluates, time in each function
# $by.total - divides time spent per function by total run time
# $by.self - same as by.total but first subtracts time spent in function above in call stack. Helps target.

```
